builtin/logical/pki/path_resign_crls.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return logical.ErrorResponse("invalid value for %s: %v", nextUpdateParam, err), nil
+	}
+
+	if nextUpdateOffset <= 0 {
+		return logical.ErrorResponse("%s parameter must be greater than 0", nextUpdateParam), nil
+	}
+
+	if crlNumber < 0 {
+		return logical.ErrorResponse("%s parameter must be 0 or greater", crlNumberParam), nil
+	}
+	if deltaCrlBaseNumber < -1 {
+		return logical.ErrorResponse("%s parameter must be -1 or greater", deltaCrlBaseNumberParam), nil
+	}
+
+	if issuerRef == "" {
+		return logical.ErrorResponse("%s parameter cannot be blank", issuerRefParam), nil
+	}
+
+	providedCrls, err := decodePemCrls(rawCrls)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	sc := b.makeStorageContext(ctx, request.Storage)
+	caBundle, err := getCaBundle(sc, issuerRef)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	if err := verifyCrlsAreFromIssuersKey(caBundle.Certificate, providedCrls); err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	revokedCerts, warnings, err := getAllRevokedCerts(providedCrls)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	now := time.Now()
+	template := &x509.RevocationList{
+		SignatureAlgorithm:  caBundle.RevocationSigAlg,
+		RevokedCertificates: revokedCerts,
+		Number:              big.NewInt(int64(crlNumber)),
+		ThisUpdate:          now,
+		NextUpdate:          now.Add(nextUpdateOffset),
+	}
+
+	if deltaCrlBaseNumber > -1 {
+		ext, err := certutil.CreateDeltaCRLIndicatorExt(int64(deltaCrlBaseNumber))
+		if err != nil {
+			return nil, fmt.Errorf("could not create crl delta indicator extension: %v", err)
+		}
+		template.ExtraExtensions = []pkix.Extension{ext}
+	}
+
+	crlBytes, err := x509.CreateRevocationList(rand.Reader, template, caBundle.Certificate, caBundle.PrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating new CRL: %w", err)
+	}
+
+	body := encodeResponse(crlBytes, format == "der")
+
+	return &logical.Response{
+		Warnings: warnings,
+		Data: map[string]interface{}{
+			"crl": body,
+		},
+	}, nil
+}
+
+func verifyCrlsAreFromIssuersKey(caCert *x509.Certificate, crls []*x509.RevocationList) error {
+	for i, crl := range crls {
+		// At this point we assume if the issuer's key signed the CRL that is a good enough check
+		// to validate that we owned/generated the provided CRL.
+		if err := crl.CheckSignatureFrom(caCert); err != nil {
+			return fmt.Errorf("CRL index: %d was not signed by requested issuer", i)
+		}
+	}
+
+	return nil
+}
+
+func encodeResponse(crlBytes []byte, derFormatRequested bool) string {
+	if derFormatRequested {
+		return base64.StdEncoding.EncodeToString(crlBytes)
+	}
+
+	block := pem.Block{
+		Type:  "X509 CRL",
+		Bytes: crlBytes,
+	}
+	// This is convoluted on purpose to ensure that we don't have trailing
+	// newlines via various paths
+	return strings.TrimSpace(string(pem.EncodeToMemory(&block)))
```

Comment:
TBH for new APIs, I'd rather leave the trailing newline, but that's me.

Reply:
Understood, I'd rather be consistent across different API returns for the same thing if possible, but I won't die on that hill either ;)

Reply:
Eh, :shrug: at some point I think I'd prefer to fix up the PEM issue that's been open forever, but...

Reply:
Further discussions with @cipherboy he brought to my attention this causes us to violate RFC7468, so will fix.

Reply:
Fixed